### ISSUE TITLE
Numerical errors in symmetric tensors lead to eventual instability

### ISSUE THREAD

**Issue by Zenome84** [tensorflow/tensorflow#19750]
I use `B = tf.matmul(A, A, transpose_b=True)` throughout my model, but I keep getting output tensors that are not symmetric. At first this is and unnoticeable amount, but after sufficient optimization steps calculations like `tf.linalg.logdet` and `tf.linalg.inv` blow up - not invertable or no cholesky decomposition.

After many steps of testing, I noticed that this happened because the input tensors were not symmetric, when they should be.

To test this out, I tried `tf.reduce_sum(tf.cast(B != tf.transpose(B), tf.int16))` to check at what point the data stops being symmetric, and this happens right at the first step.

Is this a known issue? Any recommendations?

**Comment by yaroslavvb** [comment]
It's a known issue that floating point computations can produce small errors on modern hardware, and that small errors can blow up into big errors over time of training. For instance, SSE optimized-code can lead to [slightly different results](http://blog.nag.com/2011/02/wandering-precision.html) coming out of the same computation, depending on how full the memory is.

The ultimate solution is to make your computations robust to small errors/non-determinism. But also, if you narrow down the problem, it might be possible the fix the errors. IE, non-determinism in reduce_sum has been fixed on GPU recently.

In your case, loss of symmetry implies that result of `a b'` and `a' b` are not the same. Can isolate a specific `a` and `b` that produce this?

**Comment by Zenome84** [comment]
OK, so I created a dummy script to reproduce the issue. And I apologize, I don't think the issue ever arises with `tf.matmul(A, A, transpose_b=True)` applied as such. Rather, it happens when `tf.matmul` is applied iteratively on a positive-definite matrix - i.e. `tf.matmul(tf.matmul(c, A), c, transpose_b=True)`.

Here is an example:


    import tensorflow as tf
    import numpy as np

    dim = 3
    dimN = 5

    # We will load a positive definite matrix for K
    K_init = tf.placeholder(tf.float32, [dim, dim])
    K_ = tf.Variable(K_init, trainable=False, collections=[])

    C = tf.Variable(tf.ones([dimN, dim]), dtype=tf.float32)
    a = tf.Variable(tf.ones([dimN]), dtype=tf.float32)

    # We build a Recurrent Network for K
    K = [K_]
    # Kca should be symmetric: C*K*C' is symmetric, and A is diagonal; so their sum is symmetric 
    Kca = tf.expand_dims(tf.matmul(tf.matmul(C, K[0]), C, transpose_b=True) + tf.diag(a), 0)
    # invKca should be symmetric as the inverse of a Kca
    invKca = tf.expand_dims(tf.linalg.inv(Kca[0]), 0)
    # K[1] should be symmetric since, again C'*invKca*C is symmetric.
    K = tf.concat([K, tf.expand_dims(tf.matmul(tf.matmul(C, invKca[0], transpose_a=True), C), 0)], 0)

    # Some step cost function
    costN = tf.expand_dims(tf.linalg.logdet(Kca[0]) + tf.matmul(tf.matmul([a], invKca[0]), [a], transpose_b=True)[0][0], 0)

    # The recursion
    for n in range(1, 10):
        Kca = tf.concat([Kca, tf.expand_dims(tf.matmul(tf.matmul(C, K[n]), C, transpose_b=True) + tf.diag(a), 0)], 0)
        invKca = tf.concat([invKca, tf.expand_dims(tf.linalg.inv(Kca[n]), 0)], 0)
        K = tf.concat([K, tf.expand_dims(tf.matmul(tf.matmul(C, invKca[n], transpose_a=True), C), 0)], 0)

        costN = tf.concat([costN, tf.expand_dims(tf.linalg.logdet(Kca[n]) + tf.matmul(tf.matmul([a], invKca[n]), [a], transpose_b=True)[0][0], 0)], 0)

    cost = tf.reduce_sum(costN)
    # If this example blows up, decrease the learning rate.
    step = tf.train.AdamOptimizer(0.001).minimize(cost)

    sess = tf.Session()
    init = tf.global_variables_initializer()
    sess.run(init)

    # Positive definite matrix K0
    L0 = np.array([[1.0, 0.0, 0.0], \
        [3.5, 2.5, 0.0], \
        [-2.0, -1.2, 3.0]])
    K0 = np.matmul(L0, L0.transpose())

    sess.run(K_.initializer, feed_dict={K_init:K0})
    for n in range(100):
        sess.run(step)
        print("Cost: " + str(sess.run(cost)))

    # Lo and Behold, K[10] is not symmetric
    print(sess.run(K[10]) == sess.run(K[10]).transpose())
    print(sess.run(K[10]))
    print("Done")

Now I can apply a manual fix using an intermediate step:
`B = tf.matmul(tf.matmul(c, A), c, transpose_b=True)`
`C = 0.5(B + tf.transpose(B))`

But this type of process is so common - i.e. estimates of the second moments - that I would expect a more direct implementation.  Suffice it to say that for long recursions and many dimensions, having to add this intermediate step before every `logdet` and `inv` adds a significantly noticeable difference in step time.

**Comment by Zenome84** [comment]
Also, the following variant using a custom `tensor_product` and using `tf.tensordot` suffers the same problem (this version would be necessary to work with higher moments):

    import tensorflow as tf
    import numpy as np

    #  Tensor Product for Ease
    def tensor_product(*e):
        """ Tensor product of elements """
        if len(e) == 1:
            return e
        elif len(e) == 2:
            a, b = e
            r_a = len(a.get_shape().as_list())
            r_b = len(b.get_shape().as_list())
            s_a = tf.concat([tf.shape(a), tf.constant([1] * r_b)], axis=0)
            s_b = tf.concat([tf.constant([1] * r_a), tf.shape(b)], axis=0)
            a_reshaped = tf.reshape(a, s_a)
            b_reshaped = tf.reshape(b, s_b)
            return a_reshaped * b_reshaped
        prod = e[0]
        for tensor in e[1:]:
            prod = tensor_product(prod, tensor)
        return prod

    dim = 3
    dimN = 5

    # We will load a positive definite matrix for K
    K_init = tf.placeholder(tf.float32, [dim, dim])
    K_ = tf.Variable(K_init, trainable=False, collections=[])

    C = tf.Variable(tf.ones([dimN, dim]), dtype=tf.float32)
    a = tf.Variable(tf.ones([dimN]), dtype=tf.float32)

    CC = tensor_product(C, C)
    aa = tensor_product(a, a)
    # We build a Recurrent Network for K
    K = [K_]
    # Kca should be symmetric: C*K*C' is symmetric, and A is diagonal; so their sum is symmetric 
    Kca = tf.expand_dims(tf.tensordot(CC, K[0], [[1, 3], [0, 1]]) + tf.diag(a), 0)
    # invKca should be symmetric as the inverse of a Kca
    invKca = tf.expand_dims(tf.linalg.inv(Kca[0]), 0)
    # K[1] should be symmetric since, again C'*invKca*C is symmetric.
    K = tf.concat([K, tf.expand_dims(tf.tensordot(CC, invKca[0], [[0, 2], [0, 1]]), 0)], 0)

    # Some step cost function
    costN = tf.expand_dims(tf.linalg.logdet(Kca[0]) + tf.matmul(tf.matmul([a], invKca[0]), [a], transpose_b=True)[0][0], 0)

    # The recursion
    for n in range(1, 10):
        Kca = tf.concat([Kca, tf.expand_dims(tf.tensordot(CC, K[n], [[1, 3], [0, 1]]) + tf.diag(a), 0)], 0)
        invKca = tf.concat([invKca, tf.expand_dims(tf.linalg.inv(Kca[n]), 0)], 0)
        K = tf.concat([K, tf.expand_dims(tf.tensordot(CC, invKca[n], [[0, 2], [0, 1]]), 0)], 0)

        costN = tf.concat([costN, tf.expand_dims(tf.linalg.logdet(Kca[n]) + tf.matmul(tf.matmul([a], invKca[n]), [a], transpose_b=True)[0][0], 0)], 0)

    cost = tf.reduce_sum(costN)
    # If this example blows up, decrease the learning rate.
    step = tf.train.AdamOptimizer(0.001).minimize(cost)

    sess = tf.Session()
    init = tf.global_variables_initializer()
    sess.run(init)

    # Positive definite matrix K0
    L0 = np.array([[1.0, 0.0, 0.0], \
        [3.5, 2.5, 0.0], \
        [-2.0, -1.2, 3.0]])
    K0 = np.matmul(L0, L0.transpose())

    sess.run(K_.initializer, feed_dict={K_init:K0})
    for n in range(100):
        sess.run(step)
        print("Cost: " + str(sess.run(cost)))

    # Lo and Behold, K[10] is not symmetric
    print(sess.run(K[10]) == sess.run(K[10]).transpose())
    print(sess.run(K[10]))
    print("Done")

**Comment by Zenome84** [comment]
On a final note, the quick fix `(cAc'+cA'c')/2` is not bullet-proof when we require `cAc'` to stay positive-definite. The only way I know to do this is to run a cholesky decomposition on `A` and then to norm `cA`. However, this is an even more computationally intensive procedure.

**Comment by yaroslavvb** [comment]
Not clear that this is a bug since some floating point imprecision is expected. IMHO, further isolation is required.

**Comment by rohan100jain** [comment]
This is expected behavior as Yaroslav mentioned. So I'm not sure whether there is anything to be done here.
